### DISCUSSION TITLE
Centering callout contents and/or titles

### DISCUSSION THREAD

**Discussion by cgadal** [quarto-dev/quarto-cli#4508] [Q&A]
By default, the content of callouts, and their titles, are aligned to the left.

How should I proceed to have:

- left-aligned titles, and centered content ?
- centered titles and left-aligned content ?
- both centered ?

Thanks !

**Comment by mcanouil** [comment]
For `html` format, you can tweak the CSS.
Depending on what you want you have to modify different CSS classes.

Below I set everything centered for any kind of callout.

```qmd
---
title: "Untitled"
format: html
include-in-header:
  - text: |
      <style>
      .callout.callout-style-default {
        text-align: center;
      }
      </style>
---

::: {.callout-important}
Something important
:::
```

<img width="1169" alt="image" src="https://user-images.githubusercontent.com/8896044/221033294-9b843f49-0faf-40c0-bad1-7ff99719ed8f.png">

If you want to do more custom things, render your document, open it with your favourite browser and use its developper mode (usually, right click and "inspect" menu is available).
See an example
<img width="1504" alt="image" src="https://user-images.githubusercontent.com/8896044/221033594-58ef8b16-7465-4fe5-b566-9511d2739be3.png">

For `pdf` format, you can add `keep-tex: true` in your YAML, then open the intermediary tex file.
From there, you'll have to search on the internet what you need from LaTeX to do what you want.
You might need to do the processing with a Lua filter.
There is documentation about Lua in Quarto's documentation and elsewhere on the internet.

**Reply by cgadal** [reply]
Thanks for your tips ! I am doing html presentations, and I managed to modify the whole class from a custom scss, like you suggested in the header.

However, this applies modifications for the whole document. I would you like to be able to chose different possible alignments for each call-out, but I haven't been able to do it using simple inline code :

```
::: {.callout-tip style="text-align:center;"}
Test
:::
```

**Reply by cscheid** [reply]
Try this:

```
::: {style="text-align:center;"}

::: {.callout-tip}
Test
:::

:::
```

<img width="971" alt="image" src="https://user-images.githubusercontent.com/285675/221047671-d1dc9d00-2f18-46fa-acb4-22da2803f858.png">

We should be nicer about forwarding attributes from the callout div, I'll fix that right away.

**Reply by cscheid** [reply]
That's now fixed on main:

````
---
title: "4508"
---

::: {.callout-tip style="text-align:center;"}
Test
:::
````

produces this:

<img width="985" alt="image" src="https://user-images.githubusercontent.com/285675/221049498-b4d219c8-6772-4f42-a962-9f0bb3fd4869.png">

**Reply by cgadal** [reply]
Thanks ! A side question, is there a way to center only title or only content ?

**Reply by cgadal** [reply]
> Try this:
> 
> ```
> ::: {style="text-align:center;"}
> 
> ::: {.callout-tip}
> Test
> :::
> 
> :::
> ```
> 

This does not work if using the callout with `appearance="minimal"` !

```
  ::: {style="text-align:center;"}
  
  ::: {.callout-tip appearance="minimal"}
  Test
  :::
  
  :::
```

**Reply by mcanouil** [reply]
> Thanks ! A side question, is there a way to center only title or only content ?

See my reply for advanced customisation.

or use the same trick as Carlos showed, _i.e._, using fenced fivs
```qmd
---
title: "Untitled"
format: html
---

::: {.callout-important}
:::: {style="text-align: center;"}
Something important
::::
:::
```

And now, with the modification he did
```qmd
---
title: "Untitled"
format: html
---

::: {.callout-important style="text-align:center;"}
:::: {style="text-align: left;"}
Something important
::::
:::
```

**Reply by cgadal** [reply]
Indeed. I can therefore confirm that this does not work with `appearance=minimal`.
I had that keyword when I tried it at first, hence I thought it was not working

**Reply by mcanouil** [reply]
What exactly "does not work" with `appearance=minimal`?
I do not see any issues.

### With 1.2
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/221159095-96b9ce9d-93bb-4af9-9cb9-aa61a2d61c68.png">

### With dev
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/221159456-ceecd8e5-55cd-4fb9-91b8-c203963bda48.png">

**Reply by cgadal** [reply]
- Quarto v. 1.2.335

![image](https://user-images.githubusercontent.com/48522112/221169494-fda842f1-7a42-4a80-a83f-569aa364f81c.png)

Note that compared to @cscheid  answer, the title is also not centered here, even without `appearance=minimal' ?

**Reply by mcanouil** [reply]
Oh, the issue is in `revealjs` format, not in `html`, most likely because `revealjs` css is kind of independent.
@Cgadal This is why a fully reproducible example is better, _i.e._, in this case we would have seen the issue happening in `revealjs` sooner.

**Reply by mcanouil** [reply]
@cscheid I made few testing for `revealjs` for you to possibly apply the changes you made for `html`.

The following test was done using Quarto at cc03e366904182d2c5f003d69f9f8a59f2d40eb3

```qmd
---
format: revealjs
---

## Default Appearance

::: {.callout-important}
### As expected

:::: {style="text-align: center;"}
As expected
::::
:::

::: {.callout-important style="text-align: center;"}
### Not as expected

Not as expected
:::

:::: {style="text-align: center;"}

::: {.callout-important}
### Not as expected

As expected
:::
::::

## Minimal Appearance

::: {.callout-important appearance="minimal"}
### As expected

:::: {style="text-align: center;"}
Not as expected
::::
:::

::: {.callout-important appearance="minimal" style="text-align: center;"}
### Not as expected

Not as expected
:::

:::: {style="text-align: center;"}

::: {.callout-important appearance="minimal"}
### Not as expected

Not as expected
:::
::::
```

<img width="728" alt="image" src="https://user-images.githubusercontent.com/8896044/221173226-c3692e65-b83c-468b-8b55-0f8aa7c20951.png">
<img width="727" alt="image" src="https://user-images.githubusercontent.com/8896044/221173761-5f0cb9b2-7e7a-45b4-9842-ef204a7af20a.png">

**Reply by cscheid** [reply]
To center only the title, you'll have to investigate the CSS generated by the callout and create selectors that target the specific items. You won't be able to use divs because we pull out the contents of the heading and build a separate title. (That's a requirement, since callouts can be built programmatically using `quarto.Callout` in Lua, in which case there's no `##...` element to call a "title")

**Reply by mcanouil** [reply]
Thus for `revealjs` the only way is the "inspect" approach I described in my original reply.
(That's usually my approach even if there are "easier" ways/shortcuts in Quarto)

**Reply by cgadal** [reply]
After some investigation, I found a few things:

- For some reason `simple` callout div as rendered as flex container, while `default` callouts are not, hence the difference in behaviour:

```
    .callout.callout-style-simple {
      padding: 0em 0.5em;
      border-left: solid #acacac .3rem;
      border-right: solid 1px silver;
      border-top: solid 1px silver;
      border-bottom: solid 1px silver;
      display: flex;  <-------- HERE
    }

    .callout.callout-style-default {
      border-left: solid #acacac .3rem;
      border-right: solid 1px silver;
      border-top: solid 1px silver;
      border-bottom: solid 1px silver;
    }
```

Commenting `display: flex;` of ` .callout.callout-style-simple` makes them behave the same way.  

Should this be fixed ?
Also rendering in `format: html` does not lead to `display: flex`.

I am still unsure of why the titles are not centered though

**Reply by cgadal** [reply]
@cscheid  @mcanouil I'd like to re-open this issue since I think this has not been fixed ?

Also, building on my previous answer, I am not sure how to override the default behavior:

```
---
format:
  revealjs:
    theme: [black]
    include-in-header:
      - text: |
          <style>
          .callout.callout-style-simple {
            text-align: center;
            display: inline-flexbox;
          }
          </style>
---

## test

::: {style='text-align:center;'}
:::{.callout-note appearance='minimal'}
##### Title

content

:::
:::

::: {style='text-align:center;'}
:::{.callout-note}
##### Title

content

:::
:::
```

does not change the fact that simple callouts have `display:flex`.

**Reply by mcanouil** [reply]
The following should do it:

````qmd
---
format:
  revealjs:
    include-in-header:
      - text: |
          <style>
          .callout.callout-style-simple {
            display: block;
          }
          </style>
---

## Callouts

:::: {layout-ncol="2"}

::::: {}

::: {.callout-important}
### Title

:::: {style="text-align: center;"}
Content
::::
:::

::: {.callout-important style="text-align: center;"}
### Title

Content
:::

:::: {style="text-align: center;"}

::: {.callout-important}
### Title

Content
:::
::::

:::::

::::: {}

::: {.callout-important appearance="minimal"}
### Title

:::: {style="text-align: center;"}
Content
::::
:::

::: {.callout-important appearance="minimal" style="text-align: center;"}
### Title

Content
:::

:::: {style="text-align: center;"}

::: {.callout-important appearance="minimal"}
### Title

Content
:::
::::

:::::
::::::

````

@cderv do you think there is a change to be made in `revealjs` css as the one above?

**Reply by cderv** [reply]
Regarding previous reported issue
> For some reason simple callout div as rendered as flex container, while default callouts are not, hence the difference in behaviour:

I don't know why this is `display: flex` for simple style callout and not others. 

> Also rendering in format: html does not lead to display: flex.

Callout CSS for revealjs is specific because it does not have a bootstrap dependency whereas for `format: html` callouts leverages bootstrap style. 

> does not change the fact that simple callouts have display:flex.

is `display: inline-flexbox;` valid ? My devtools console says it is not and I can find it on MZN . Probably using a valid value will work like in @mcanouil example

> do you think there is a change to be made in revealjs css as the one above?

If we think this `display: flex` has indeed no specific meaning for simple callout, then yes we could remove it. I don't know why this style is set... 🤔
